tests/test_util_matcher.py

```diff
@@ -0,0 +1,259 @@
+        matcher.should_store_similarity_matrix = True
+
+        # There are undefeatable matches for the first attribute / weight here, and so
+        # by default the algorithm will not even check the second attribute.
+        a = [('a', -8), ('very clear', 0), ('way', 33), ('forward', 2)]
+        b = [('very clear', 33), ('forward', -1), ('a', 5), ('way', 2)]
+
+        assert matcher.get_indices(a, b) == [2, 0, 3, 1]
+        partial_matrix = matcher.similarity_matrix
+
+        matcher.should_go_through_every_attribute = True
+
+        # Result definitely shouldn't change
+        assert matcher.get_indices(a, b) == [2, 0, 3, 1], formatted_matrix(matcher)
+
+        # But in this case the matrix should have
+        assert matcher.similarity_matrix != partial_matrix, formatted_matrix(matcher)
+
+
+def formatted_matrix(matcher: ObjectListMatcher) -> str:
+    lines = ['<Similarity Matrix']
+    for b_similarities in matcher.similarity_matrix:
+        l = ''
+        for b_sim in b_similarities:
+            l += f'{b_sim:1.4f}  '
+        lines.append(l)
+    return '\n'.join(lines) + '\n>'
+
+
+class TMatchClassFields(TestCase):
+    def test_matching_works(self):
+        a, b = self._get_car_lists()
+
+        attr_to_weight = {(lambda c: c.seats): 4, (lambda c: c.name): 1,
+                          (lambda c: c.features): 5}
+        matcher = ObjectListMatcher(attr_to_weight)
+
+        assert matcher.get_indices(a, b) == [2, None, 1, 0]
+
+    def _get_car_lists(self):
+        a = [Car(1, 'Speedy', ['gps', 'heater']), Car(2, 'Cheaporghiny', []),
+             Car(16, 'Half-a-Bus', ['buttons']), Car(5, 'Normal Model 1', ['music'])]
+        b = [Car(3, 'Mödel V5', ['gps']), Car(19, 'Cyberbus', ['buttons']),
+             Car(2, 'Sheeporghiny', ['gps', 'heater', 'sheep sound button'])]
+        return a, b
+
+    def test_dominating_name_weights(self):
+        a, b = self._get_car_lists()
+        attr_to_weight = {(lambda c: c.seats): 0.5, (lambda c: c.name): 9,
+                          (lambda c: c.features): 1.2}
+
+        matcher = ObjectListMatcher(attr_to_weight)
+
+        assert matcher.get_indices(a, b) == [None, 2, 1, 0]
+
+    def test_minimum_similarity(self):
+        a, b = self._get_car_lists()
+
+        attr_to_weight = {(lambda c: c.seats): 3, (lambda c: c.features): 8}
+        matcher = ObjectListMatcher(attr_to_weight)
+        matcher.should_store_similarity_matrix = True
+
+        assert matcher.get_indices(a, b) == [2, 0, 1, None], formatted_matrix(matcher)
+
+        matcher.minimum_similarity_ratio = 0.71
+        assert matcher.get_indices(a, b) == [2, None, 1, None], formatted_matrix(
+            matcher)
+
+        matcher.minimum_similarity_ratio = 0.9
+        assert matcher.get_indices(a, b) == [None, None, 1, None], formatted_matrix(
+            matcher)
+
+
+class Car:
+    def __init__(self, seats: int, name: str, features: List[str]):
```

Comment:
`@dataclass` is good for this stuff BTW

Reply:
Oh.. somehow I thought they were added in 3.8. Will replace.